### PR TITLE
feat(responses): bridge /v1/responses to non-OpenAI providers (#825)

### PR DESCRIPTION
**Pull request by jarvis9443**
## Problem

Codex (and any OpenAI Responses-API client) pointed at a non-OpenAI model on `POST /v1/responses` got a hard `400`:

```
{"error":{"message":"request payload is invalid: model `opus-4.7` is not an OpenAI provider; /v1/responses requires OpenAI","type":"invalid_request_error"}}
```

The handler was OpenAI-only — it rejected every non-OpenAI provider before dispatch. codex-tui speaks only the Responses API, so pointing it at an Anthropic model was impossible.

## Fix

Bridge `/v1/responses` to non-OpenAI providers, mirroring how `/v1/messages` already does cross-provider:

- **OpenAI models**: unchanged — body forwarded verbatim to the upstream `/v1/responses`.
- **Non-OpenAI models**: translate the Responses request into the canonical `ChatFormat`, dispatch through the provider `Bridge` (the same path `/v1/chat/completions` uses), and re-encode the reply back into the Responses shape — non-streaming JSON and streaming SSE (`response.created` → `response.output_item.added` → `response.output_text.delta` / `response.function_call_arguments.delta` → `response.completed`).

A Model Group can mix OpenAI and non-OpenAI targets and fail over across both kinds. Usage events are emitted on both paths (non-streaming from the handler, streaming from the response stream's Drop guard), with Anthropic cache counters carried through.

New module `crates/aisix-proxy/src/responses_bridge.rs` holds the Responses↔ChatFormat request/response/SSE translation.

### Multi-turn tool loops

For codex's agent loop to work beyond a single text turn, the Anthropic bridge's `split_system` also needed two fixes (it previously affected `/v1/chat/completions`→Anthropic too):

- translate an assistant turn's `tool_calls` into Anthropic `tool_use` blocks (previously dropped, which orphaned the following `tool_result`)
- merge consecutive same-role turns so the alternating-role invariant Anthropic enforces holds for parallel tool calls and tool-result replays

### Not bridged

OpenAI-only Responses knobs with no cross-provider equivalent are dropped rather than forwarded (forwarding them would 400 a backend like Anthropic): `reasoning`, `store`/`previous_response_id`, hosted tools (only `type:"function"` translates), `text`/`metadata`/`service_tier`. The OpenAI verbatim path is unaffected.

## Behavior change

`/v1/responses` against a non-OpenAI provider no longer returns `400`; it is bridged and returns a `200` Responses-shape reply. The pre-existing `responses-endpoint-e2e` cases that asserted the old `400` boundary are updated to the new contract.

## Tests

- `responses_bridge` unit tests: request/response/SSE translation, parallel-tool folding, withheld-completed-until-usage.
- `responses.rs` integration tests: non-OpenAI model bridges to Responses shape (non-streaming + streaming) against a mock Anthropic upstream.
- `split_system` unit tests: assistant `tool_calls`→`tool_use`, parallel tool-result merging.
- DP E2E `responses-cross-provider-e2e`: non-streaming, streaming (the codex repro), and a multi-turn tool loop against a mock Anthropic upstream, asserting Responses output, SSE events, tool_use/tool_result translation, and usage events.

## Docs

`docs/integration/responses.md` rewritten to document cross-provider support, the bridge behavior, and the unbridged fields.

Fixes api7/AISIX-Cloud#825

## Release Notes

* **New Features**
  * Added cross-provider support for `POST /v1/responses`, including non-OpenAI backends with Responses-shaped streaming, usage reporting (including cache token counters), and tool-calling/tool-loop support.

* **Bug Fixes**
  * Improved Anthropic tool-call translation (no empty message content) and enforced correct role alternation during tool loops.
  * Fixed streaming event ordering so `response.completed` appears after final usage.

* **Documentation**
  * Updated `/v1/responses` docs with cross-provider behavior and noted dropped OpenAI-only fields.

* **Tests**
  * Expanded E2E coverage for bridging, streaming event sequences, tool loops, and completion/usage edge cases.

Fixes api7/AISIX-Cloud#788

### CODE REVIEW

File: crates/aisix-proxy/src/responses_bridge.rs [hunk 1]
```diff
@@ -0,0 +1,1331 @@
...
+        .and_then(|v| v.as_f64())
+        .map(|f| f as f32);
+    chat.top_p = body.get("top_p").and_then(|v| v.as_f64()).map(|f| f as f32);
+    // Responses calls the cap `max_output_tokens`; tolerate `max_tokens`
+    // too for clients that send the chat-style name.
+    chat.max_tokens = body
+        .get("max_output_tokens")
+        .or_else(|| body.get("max_tokens"))
+        .and_then(|v| v.as_u64())
+        .map(|n| n as u32);
```
[the remark this answers is not in the document]
Comment (jarvis9443) [reply crates/aisix-proxy/src/responses_bridge.rs]: Done — now uses `u32::try_from(n).ok()`, so an out-of-range `max_output_tokens` is dropped (left unset) rather than wrapping to a small/zero cap. Unit test added.

File: crates/aisix-proxy/src/responses_bridge.rs [hunk 2]
```diff
@@ -0,0 +1,1331 @@
...
+/// Per-tool-call streaming state.
+#[derive(Debug)]
+struct ToolCallState {
+    item_id: String,
+    call_id: String,
+    name: String,
+    output_index: u32,
+    arguments: String,
+    item_added: bool,
+}
```
[the remark this answers is not in the document]
Comment (jarvis9443) [reply crates/aisix-proxy/src/responses_bridge.rs:404]: Good catch. `close_items` is now idempotent (a single `items_closed` guard), so the finish chunk plus a later `force_finish` (when completed was withheld for usage) no longer double-emit the `*.done` events. Regression test added.

File: crates/aisix-proxy/src/responses_bridge.rs [hunk 3]
```diff
@@ -0,0 +1,1331 @@
...
+    tool_calls: std::collections::BTreeMap<u64, ToolCallState>,
+    /// Withheld terminal status while waiting on a trailing usage frame.
+    pending_status: Option<&'static str>,
+    // Accumulated usage (max semantics, robust to double-emit).
+    usage_seen: bool,
+    prompt_tokens: u32,
+    completion_tokens: u32,
+    reasoning_tokens: u32,
+    cached_prompt_tokens: u32,
+    cache_read_tokens: u32,
```
[the remark this answers is not in the document]
Comment (jarvis9443) [reply crates/aisix-proxy/src/responses_bridge.rs:438]: Fixed. The encoder accumulates `total_tokens` now, and `responses_usage_json` keeps a provider-supplied total when present (falls back to prompt+completion only when it's 0). Test added.

File: crates/aisix-proxy/src/responses_bridge.rs [hunk 4]
```diff
@@ -0,0 +1,1331 @@
...
+        // ── Finish ──
+        if let Some(fr) = chunk.finish_reason.as_ref() {
+            events.extend(self.close_items());
+            let (status, _) = responses_status(fr);
+            if self.usage_seen {
+                events.push(self.completed_event(status));
+            } else {
+                // Hold response.completed until the trailing usage frame.
+                self.pending_status = Some(status);
+            }
```
[the remark this answers is not in the document]
Comment (jarvis9443) [reply crates/aisix-proxy/src/responses_bridge.rs:756]: Fixed. The streaming terminal event now carries `incomplete_details.reason` (threaded through `pending_reason` into `completed_event`), matching the non-streaming JSON path. Test added.

File: crates/aisix-proxy/src/responses_bridge.rs [hunk 5]
```diff
@@ -0,0 +1,1331 @@
...
+        // on a block, emit a terminal error event rather than completing.
+        if let Some(chain) = output_guardrail.as_ref() {
+            if !content_text.is_empty() || !tool_call_fragments.is_empty() {
+                let mut message =
+                    aisix_gateway::ChatMessage::assistant(std::mem::take(&mut content_text));
+                if !tool_call_fragments.is_empty() {
+                    message.extra.insert(
+                        "tool_calls".to_string(),
+                        Value::Array(std::mem::take(&mut tool_call_fragments)),
+                    );
```
[the remark this answers is not in the document]
Comment (jarvis9443) [reply crates/aisix-proxy/src/responses_bridge.rs]: Addressed together with the buffering change below: when an output guardrail is attached the bridge now scans the encoder's fully-reassembled assistant output — canonical `{id,type,function:{name,arguments}}` tool calls and full text — instead of raw deltas, so a literal split across argument deltas is scanned as one string.

File: crates/aisix-proxy/src/responses.rs [hunk 6]
```diff
@@ -942,6 +964,259 @@ async fn responses_to_target(
...
+                futures::stream::once(std::future::ready(Ok::<_, aisix_gateway::BridgeError>(
+                    first_chunk,
+                )))
+                .chain(upstream),
+            )
+        } else {
+            upstream
+        };
+        state.health.record_success(requested_model);
+        state.runtime_status.mark_healthy(model_id);
```
[the remark this answers is not in the document]
Comment (jarvis9443) [reply crates/aisix-proxy/src/responses.rs]: Fixed. The streaming branch now records health against `model.display_name` (the resolved target), matching the non-streaming branch.

File: crates/aisix-proxy/src/responses.rs [hunk 7]
```diff
@@ -942,6 +964,259 @@ async fn responses_to_target(
...
+        let requested_model_c = requested_model.to_string();
+        let api_key_id_c = api_key_id.to_string();
+        let client_c = client_ctx.clone();
+        let attempt_c = attempt.clone();
+        let sse_body = crate::responses_bridge::build_responses_bridge_stream(
+            upstream,
+            encoder,
+            started,
+            output_guardrail,
+            requested_model.to_string(),
```
[the remark this answers is not in the document]
Comment (jarvis9443) [reply crates/aisix-proxy/src/responses.rs]: Fixed. When an output guardrail is attached the bridge now holds the encoded SSE back and releases it only after the end-of-stream scan passes, with a capped buffer that fails closed on overflow — mirroring the verbatim path's #719 BufferFull default. A streaming response's 200 is already committed by the first-chunk failover peek, so a block surfaces as a terminal `error` event rather than a 422, but no output_text/tool delta reaches the client. Added integration tests asserting the blocked literal never appears in the stream and that a clean stream is released in full.

File: crates/aisix-proxy/src/responses.rs [hunk 8]
```diff
@@ -942,6 +964,259 @@ async fn responses_to_target(
...
+            move |comp| {
+                // Streams that reach here are committed 200s.
+                let usage = ResponseUsage {
+                    prompt_tokens: comp.prompt_tokens,
+                    completion_tokens: comp.completion_tokens,
+                    reasoning_tokens: comp.reasoning_tokens,
+                    cached_prompt_tokens: comp.cached_prompt_tokens,
+                    cache_creation_tokens: 0,
+                    cache_read_tokens: comp.cache_read_tokens,
+                };
```
[the remark this answers is not in the document]
Comment (jarvis9443) [reply crates/aisix-proxy/src/responses.rs:1114]: Fixed. `ResponsesStreamCompletion` now accumulates `cache_creation_tokens` from the chunk usage and the streaming usage event maps it (no longer hard-coded 0).

File: tests/e2e/src/cases/responses-cross-provider-e2e.test.ts [hunk 9]
```diff
@@ -0,0 +1,386 @@
...
+                attrs[a.key] =
+                  v.stringValue ?? String(v.intValue ?? v.boolValue ?? "");
+              }
+              spanAttrs.push(attrs);
+            }
+          }
+        }
+      } catch {
+        // ignore malformed bodies — assertions fail on missing spans
+      }
```
[the remark this answers is not in the document]
Comment (jarvis9443) [reply tests/e2e/src/cases/responses-cross-provider-e2e.test.ts]: Done — the OTLP receiver now logs parse errors via `console.error`.

File: crates/aisix-proxy/src/responses_bridge.rs [hunk 10]
```diff
@@ -0,0 +1,1517 @@
...
+                aisix_guardrails::Guardrail::check_output(chain.as_ref(), &synth).await
+            {
+                tracing::warn!(
+                    guardrail_hook = "output",
+                    model = %model_label,
+                    reason = %reason,
+                    "guardrail blocked streaming /v1/responses (cross-provider) response",
+                );
+                yield Ok(bytes::Bytes::from(guardrail_error_frame(guardrail_name.as_deref())));
+                return;
```
[the remark this answers is not in the document]
Comment (jarvis9443) [reply crates/aisix-proxy/src/responses_bridge.rs:1072]: Fixed. Added a `guardrail_blocked` flag to `ResponsesStreamCompletion`, set it on both fail-closed paths (content block + buffer overflow), and the Drop-guard callback now emits the usage event as a 422 marked guardrail_blocked (still billing the upstream tokens) — matching the non-streaming path. Covered by `streaming_cross_provider_block_emits_guardrail_blocked_usage_event`.
